src/server/engine_shard_set.cc

```diff
@@ -100,34 +101,79 @@ bool EngineShard::DefragTaskState::IsRequired() {
 
 // for now this does nothing
 bool EngineShard::DoDefrag() {
-  // TODO - Impl!!
-  return defrag_state_.cursor > 0;
+  // --------------------------------------------------------------------------
+  // NOTE: This task is running with exclusive access to the shard.
+  // i.e. - Since we are using shared noting access here, and all access
+  // are done using fibers, This fiber is run only when no other fiber in the
+  // context of the controlling thread will access this shard!
+  // --------------------------------------------------------------------------
+
+  constexpr size_t kCursorCountRelation = 50;
+  const float threshold = GetFlag(FLAGS_mem_utilization_threshold);
+
+  auto& slice = db_slice();
+  DCHECK(slice.IsDbValid(kDefaultDbIndex));
+  auto [prime_table, expire_table] = slice.GetTables(kDefaultDbIndex);
+  //  otherwise we will end with negative invalid segment id - crashed at CI
+  CHECK(prime_table->depth() < 40)
```

Comment:
it's your assumption, right? Please add todo to remove this check.

Reply:
This check is based on "dash_internal.h line 679 (function segment_id). This can be removed however since I found the reason for the crash in CI. I left it in just in case, but I will remove this.

Reply:
I did not understand this part. Did you find the reason? what was the reason for the crash?

Reply:
in any case, I prefer using `DCHECK_LT` if you need it for tests